ocaml/xapi/message_forwarding.ml

```diff
@@ -3436,6 +3436,26 @@ functor
         let local_fn = Local.Host.get_sched_gran ~self in
         do_op_on ~local_fn ~__context ~host:self (fun session_id rpc ->
             Client.Host.get_sched_gran rpc session_id self)
+
+      let apply_updates ~__context ~self ~hash =
+        let uuid = host_uuid ~__context self in
+        info "Host.apply_updates: host = '%s'; hash = '%s'" uuid hash;
+        Xapi_pool_helpers.with_pool_operation
```

Comment:
You still don't exist `with_pool_operation` before the reboot or xapi restart. To make it easier, perhaps apply all guidance cases outside of this operation. And move this entire block of code into the Xapi_host function, so you don't need to worry about the return value of the API, and return unit as before (as it all runs on the master).